controllers/front/ProductController.php

```diff
@@ -83,6 +83,13 @@ public function canonicalRedirection($canonical_url = '')
         }
     }
 
+    public function getCanonicalUrl(): string
+    {
+        $product = $this->context->smarty->getTemplateVars('product');
```

Comment:
```suggestion
        $product = $this->context->smarty->getTemplateVars('product');
        if (!($product instanceof Product)) {
            return;
        }
```

Make sure the var has the good type. We should never trust and array nor an stdClass.